source/server/admin/stats_render.cc

```diff
@@ -262,5 +299,124 @@ void StatsJsonRender::collectBuckets(const std::string& name,
+  response.add(fmt::format("# TYPE {0} {1}\n", prefixed_tag_extracted_name, "gauge"));
+  for (const Stats::TextReadoutSharedPtr& metric : text_readout) {
+    auto tags = metric->tags();
+    tags.push_back(Stats::Tag{"text_value", metric->value()});
+    const std::string formTags = formattedTags(tags);
+    response.add(absl::StrCat(prefixed_tag_extracted_name, "{", formTags, "} 0\n"));
+  }
+}
+
+// Writes output for a Prometheus stat of type Histogram.
+void PrometheusStatsRender::generate(Buffer::Instance& response,
+                                     const std::string& prefixed_tag_extracted_name,
+                                     const std::vector<Stats::HistogramSharedPtr>& histogram) {
+  response.add(fmt::format("# TYPE {0} {1}\n", prefixed_tag_extracted_name, "histogram"));
+  for (const auto& metric : histogram) {
+    auto parent_histogram = dynamic_cast<Stats::ParentHistogram*>(metric.get());
+    if (parent_histogram != nullptr) {
+      // TODO(rulex123): support for the 3 histogram "bucket modes" offered in other
+      // stats formats hasn't been ported to Prometheus yet (see Utility::HistogramBucketsMode in
+      // source/server/admin/utils.h).
+      const std::string tags = formattedTags(parent_histogram->tags());
+      const std::string hist_tags = parent_histogram->tags().empty() ? EMPTY_STRING : (tags + ",");
+
+      const Stats::HistogramStatistics& stats = parent_histogram->cumulativeStatistics();
+      Stats::ConstSupportedBuckets& supported_buckets = stats.supportedBuckets();
+      const std::vector<uint64_t>& computed_buckets = stats.computedBuckets();
+      for (size_t i = 0; i < supported_buckets.size(); ++i) {
+        double bucket = supported_buckets[i];
+        uint64_t value = computed_buckets[i];
+        // We want to print the bucket in a fixed point (non-scientific) format. The fmt library
+        // doesn't have a specific modifier to format as a fixed-point value only so we use the
+        // 'g' operator which prints the number in general fixed point format or scientific format
+        // with precision 50 to round the number up to 32 significant digits in fixed point format
+        // which should cover pretty much all cases
+        response.add(absl::StrCat(prefixed_tag_extracted_name, "_bucket{", hist_tags, "le=\"",
+                                  fmt::format("{0:.32g}", bucket), "\"} ", value, "\n"));
+      }
+
+      response.add(absl::StrCat(prefixed_tag_extracted_name, "_bucket{", hist_tags, "le=\"+Inf\"} ",
+                                stats.sampleCount(), "\n"));
+      response.add(absl::StrCat(prefixed_tag_extracted_name, "_sum{", tags, "} ",
+                                fmt::format("{0:.32g}", stats.sampleSum()), "\n"));
+      response.add(absl::StrCat(prefixed_tag_extracted_name, "_count{", tags, "} ",
+                                stats.sampleCount(), "\n"));
+    }
+  }
+}
+
+void PrometheusStatsRender::finalize(Buffer::Instance&) {}
```

Comment:
Why are these 4 empty? Isn't this the StatsRender interface? At least add a comment about what's going on here.

Reply:
Yeah, I agree this is a bit odd. I will refactor the `StatsRender` interface so that we don't end up with these empty methods in the prometheus render.

Reply:
I think it's OK for 'finalize' to be empty for prom -- this is empty for 'text' mode as well. It was needed for JSON format because the structure of the way the histograms are written out makes it impossible to stream the histograms. They are to collected and then emitted all once from the finalize routine.